### PR TITLE
Orc: Support row group bloom filters

### PR DESCRIPTION
**Pull request by deadwind4**
Issue link: https://github.com/apache/iceberg/issues/5218

Add `write.orc.bloom.filter.columns` and `write.orc.bloom.filter.fpp` options.
Enable these options in ORC class.

### CODE REVIEW

**Comment by deadwind4** [comment]
cc @rdblue @RussellSpitzer @huaxingao @kbendick  Please review it, thanks a lot.

File: docs/configuration.md [hunk 1]
```diff
@@ -64,6 +64,8 @@ Iceberg tables support table properties to configure table behavior, like the de
 | write.orc.block-size-bytes         | 268435456 (256 MB) | Define the default file system block size for ORC files |
 | write.orc.compression-codec        | zlib               | ORC compression codec: zstd, lz4, lzo, zlib, snappy, none |
 | write.orc.compression-strategy     | speed              | ORC compression strategy: speed, compression |
+| write.orc.bloom.filter.columns     | (not set)          | Comma separated list of column names for which a Bloom filter must be created |
+| write.orc.bloom.filter.fpp         | 0.05               | False positive probability for Bloom filter (must > 0.0 and < 1.0) |
```
Comment (kbendick) [inline docs/configuration.md:68]: You might want to match the parquet configurations a bit more closely.

They are
```
| write.parquet.bloom-filter-enabled.column.col1          | (not set) | Enables writing a bloom filter for the column: col1|
| write.parquet.bloom-filter-max-bytes | 1048576 (1 MB)   | The maximum number of bytes for a bloom filter bitset |
```

So you could do `write.orc.bloom-filter-enabled.column.col1`. This also matches other config value formatting that's per column, such as `write.metadata.metrics.column.col1`.

For the `fpp`, as that seems to be how the value is set on the ORC bloom filter, I would suggest keeping it that way. But if the parquet implementation is translating from the max-bytes to fpp, then possibly setting the config that way for consistency (but I doubt that it is).

File: orc/src/main/java/org/apache/iceberg/orc/ORC.java [hunk 2]
```diff
@@ -220,6 +224,10 @@ private static class Context {
       private final CompressionKind compressionKind;
       private final CompressionStrategy compressionStrategy;
 
+      private final String bloomFilterColumns;
+
+      private final double bloomFilterFpp;
```
Comment (huaxingao) [inline orc/src/main/java/org/apache/iceberg/orc/ORC.java:229]: nit: shall we remove the blank lines in L226 and L228 so `bloomFilterColumns` and `bloomFilterFpp` will be grouped together with the rest of the properties?

File: orc/src/main/java/org/apache/iceberg/orc/ORC.java [hunk 3]
```diff
@@ -285,9 +305,24 @@ static Context dataContext(Map<String, String> config) {
         strategyAsString =
             PropertyUtil.propertyAsString(config, ORC_COMPRESSION_STRATEGY, strategyAsString);
         CompressionStrategy compressionStrategy = toCompressionStrategy(strategyAsString);
+        String bloomFilterColumns =
```
Comment (huaxingao) [inline orc/src/main/java/org/apache/iceberg/orc/ORC.java:308]: nit: add a blank line before this?

File: orc/src/main/java/org/apache/iceberg/orc/ORC.java [hunk 4]
```diff
@@ -285,9 +305,24 @@ static Context dataContext(Map<String, String> config) {
...
+            PropertyUtil.propertyAsString(config, OrcConf.BLOOM_FILTER_COLUMNS.getAttribute(), "");
+        bloomFilterColumns =
+            PropertyUtil.propertyAsString(config, ORC_BLOOM_FILTER_COLUMNS, bloomFilterColumns);
+
+        double bloomFilterFpp =
+            PropertyUtil.propertyAsDouble(config, OrcConf.BLOOM_FILTER_FPP.getAttribute(), 0.05);
```
Comment (huaxingao) [inline orc/src/main/java/org/apache/iceberg/orc/ORC.java:314]: Is `0.05` the default fpp for Orc? Can we define a constant for this `0.05`?

**Comment by huaxingao** [comment]
For the test, shall we also check the read path to make sure bloom filters are there?

File: orc/src/main/java/org/apache/iceberg/orc/ORC.java (hunk 2 above)
Comment (kbendick) [reply orc/src/main/java/org/apache/iceberg/orc/ORC.java:229]: +1 to that. None of the other private fields have blank lines.

**Comment by deadwind4** [comment]
> For the test, shall we also check the read path to make sure bloom filters are there?

I find the ORC SDK will auto recognize the bloom filter in the following code. Is it necessary to test the logic of ORC itself? IMO, we don't have to test it. @huaxingao 

https://github.com/apache/orc/blob/a49f87d492aa62ec2be2d4bce2fcfe1f53ca05d9/java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java#L890-L912

File: docs/configuration.md (hunk 1 above)
Comment (deadwind4) [reply docs/configuration.md:68]: IMO, in this PR, we align `write.parquet.bloom-filter-enabled.column` and `write.orc.bloom.filter.columns` with native parquet and orc options. 

We can create a new issue to unify these options. 
In the new issue, we discuss a new option of no parquet and orc tag like `write.rowgroup.bloom-filter.columns` or `write.rowgroup.bloom-filter-enabled.column` to unify file format options. 
Maybe ·write.parquet.bloom-filter-enabled.column = id, name· or `[id, name]` is more elegant than `write.rowgroup.bloom-filter.columns.id = true`.

PS: 
I find the Trino iceberg connector use the style like `orc_bloom_filter_columns = [id, name]`
https://trino.io/docs/current/connector/iceberg.html#iceberg-table-properties

I'm happy to diver this.
@kbendick What do you think?

File: orc/src/main/java/org/apache/iceberg/orc/ORC.java [hunk 5]
```diff
@@ -286,8 +305,25 @@ static Context dataContext(Map<String, String> config) {
...
+        String bloomFilterColumns =
+            PropertyUtil.propertyAsString(config, OrcConf.BLOOM_FILTER_COLUMNS.getAttribute(), "");
+        bloomFilterColumns =
+            PropertyUtil.propertyAsString(config, ORC_BLOOM_FILTER_COLUMNS, bloomFilterColumns);
+
+        double bloomFilterFpp =
+            PropertyUtil.propertyAsDouble(
+                config, OrcConf.BLOOM_FILTER_FPP.getAttribute(), BLOOM_FILTER_FPP_DEFAULT);
+        bloomFilterFpp =
+            PropertyUtil.propertyAsDouble(config, ORC_BLOOM_FILTER_FPP, bloomFilterFpp);
```
Comment (huaxingao) [inline orc/src/main/java/org/apache/iceberg/orc/ORC.java:317]: nit: Shall we add a `Preconditions.checkArgument` to make sure the `bloomFilterFpp` is `> 0.0 and < 1.0`?

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 6]
```diff
@@ -177,6 +177,12 @@ private TableProperties() {}
   public static final String AVRO_COMPRESSION_LEVEL_DEFAULT = null;
 
   public static final String ORC_STRIPE_SIZE_BYTES = "write.orc.stripe-size-bytes";
+
+  public static final String ORC_BLOOM_FILTER_COLUMNS = "write.orc.bloom.filter.columns";
+
+  public static final String ORC_BLOOM_FILTER_FPP = "write.orc.bloom.filter.fpp";
+  public static final double BLOOM_FILTER_FPP_DEFAULT = 0.05;
```
Comment (huaxingao) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:184]: nit: `BLOOM_FILTER_FPP_DEFAULT` => `ORC_BLOOM_FILTER_FPP_DEFAULT`?

**Comment by huaxingao** [comment]
LGTM. Thanks a lot @deadwind4 for working on this!
cc @kbendick @rdblue

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 7]
```diff
@@ -177,6 +177,12 @@ private TableProperties() {}
   public static final String AVRO_COMPRESSION_LEVEL_DEFAULT = null;
 
   public static final String ORC_STRIPE_SIZE_BYTES = "write.orc.stripe-size-bytes";
+
+  public static final String ORC_BLOOM_FILTER_COLUMNS = "write.orc.bloom.filter.columns";
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:181]: Nit: Ideally there would be a default value, `ORC_BLOOM_FILTER_COLUMNS_DEFAULT`. Even if it’s null (like Avro compression level default) or empty string, I think it’s better for readability and safe refactoring later to have the default defined where the rest are.

File: core/src/main/java/org/apache/iceberg/TableProperties.java (hunk 6 above)
Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/TableProperties.java:184]: +1. Other configs are format specific for their defaults, and parquet doesn’t use this configuration which could cause confusion.

File: docs/configuration.md (hunk 1 above)
Comment (kbendick) [reply docs/configuration.md:68]: I’m good with following up on anything in a new issue.

This is almost there and we should leave it as is (minus any nits). Using configs similar to those offered by native file formats is a good call imo.

**Review by kbendick — APPROVED** [review]
This LGTM, outside of the two nits on the config values.

Thank you for working on this @deadwind4!

**Comment by deadwind4** [comment]
Thank a lot @huaxingao @kbendick for reviewing this.

**Comment by deadwind4** [comment]
@kbendick Could we merge it?

**Comment by pvary** [comment]
@deadwind4: I understand, that the ORC supports these filters and tests the functions, but I would still like to see some tests validating, that the filters are there. I could envision a situation where ORC changes the way to support bloom filters, and our tests fails to recognize this.
If, and only if, there is an easy way to check it, then it would be good to have a test here as well to validate that the filters are created. 

Thanks, Peter

**Comment by deadwind4** [comment]
@pvary Thank you for your feedback. I have added a test case to validate bloom filters in ORC files and push a new commit. 
@kbendick @pvary Please review this. Thanks a lot.

**Comment by pvary** [comment]
@deadwind4: Could you please check the failed tests.

Thanks,
Peter

**Comment by deadwind4** [comment]
@pvary All checks have passed after I rebased the code and reran the CI.

**Comment by pvary** [comment]
Thanks @deadwind4 for the PR and @kbendick for the review!

**Comment by chenwyi2** [comment]
ORC_BLOOM_FILTER_COLUMNS this property will work on spark? when i set write.orc.bloom.filter.columns=xx and used spark to write data, i found that bloomfilter had no effect on querying.

**Pull request merged**